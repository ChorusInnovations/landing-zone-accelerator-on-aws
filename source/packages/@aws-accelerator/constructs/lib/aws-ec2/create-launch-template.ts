/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as fs from 'fs';

export interface ILaunchTemplateResource extends cdk.IResource {
  /**
   * The version number.
   */
  readonly version: string;
  /**
   * The ID of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateId: string;
  /**
   * The name of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateName: string;
}
export type PrivateIpAddressConfig = {
  primary: boolean | undefined;
  privateIpAddress: string | undefined;
};
export type NetworkInterfaceItemConfig = {
  associateCarrierIpAddress: boolean | undefined;
  associatePublicIpAddress: boolean | undefined;
  deleteOnTermination: boolean | undefined;
  description: string | undefined;
  deviceIndex: number | undefined;
  groups: string[] | undefined;
  interfaceType: string | undefined;
  networkCardIndex: number | undefined;
  networkInterfaceId: string | undefined;
  privateIpAddress: string | undefined;
  secondaryPrivateIpAddressCount: number | undefined;
  subnetId: string | undefined;
  privateIpAddresses: PrivateIpAddressConfig[] | undefined;
};
export type EbsProperty = {
  deleteOnTermination?: boolean;
  encrypted?: boolean;
  iops?: number;
  kmsKeyId?: string;
  snapshotId?: string;
  throughput?: number;
  volumeSize?: number;
  volumeType?: string;
};

export type BlockDeviceMappingItem = {
  deviceName: string;
  ebs?: EbsProperty;
};

export interface LaunchTemplateProps {
  /*
   * Path to user data.
   */
  readonly userData?: string;
  /*
   * VpcName
   */
  readonly vpc: string;
  /*
   * Name of Launch Template
   */
  readonly name: string;
  /*
   * Name of Application
   */
  readonly appName: string;
  readonly blockDeviceMappings?: BlockDeviceMappingItem[];
  readonly securityGroups?: string[];
  readonly networkInterfaces?: NetworkInterfaceItemConfig[];
  readonly instanceType: string;
  readonly keyPair?: string;
  readonly iamInstanceProfile?: string;
  readonly imageId: string;
  readonly enforceImdsv2?: boolean;
}

export class LaunchTemplate extends cdk.Resource implements ILaunchTemplateResource {
  public readonly version: string;
  public readonly launchTemplateId: string;
  public readonly launchTemplateName: string;

  constructor(scope: Construct, id: string, props: LaunchTemplateProps) {
    super(scope, id);

    let metadataOptions: cdk.aws_ec2.CfnLaunchTemplate.MetadataOptionsProperty;

    if (props.enforceImdsv2 === false) {
      metadataOptions = { httpTokens: 'optional' };
    } else {
      metadataOptions = { httpTokens: 'required' };
    }

    const launchTemplate = new cdk.aws_ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
      launchTemplateData: {
        blockDeviceMappings: props.blockDeviceMappings ?? undefined,
        securityGroupIds: props.securityGroups ?? undefined,
        networkInterfaces: props.networkInterfaces ?? undefined,
        instanceType: props.instanceType,
        keyName: props.keyPair ?? undefined,
        imageId: props.imageId,
        iamInstanceProfile: { name: props.iamInstanceProfile ?? undefined },
        userData: cdk.Fn.base64(fs.readFileSync(props.userData!, 'utf8')) ?? undefined,
        metadataOptions,
      },
      launchTemplateName: props.name,
    });

    this.launchTemplateId = launchTemplate.ref;
    this.launchTemplateName = launchTemplate.launchTemplateName!;
    this.version = launchTemplate.attrLatestVersionNumber;
  }
}
