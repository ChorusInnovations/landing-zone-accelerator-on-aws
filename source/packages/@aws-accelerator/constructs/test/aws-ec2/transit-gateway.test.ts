import * as cdk from 'aws-cdk-lib';
import { SynthUtils } from '@aws-cdk/assert';
import { TransitGatewayRouteTableAssociation } from '../../lib/aws-ec2/transit-gateway';

const testNamePrefix = 'Construct(TransitGatewayRouteTableAssociation): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new TransitGatewayRouteTableAssociation(stack, 'TransitGatewayRouteTableAssociation', {
  transitGatewayAttachmentId: 'transitGatewayAttachmentId',
  transitGatewayRouteTableId: 'transitGatewayRouteTableId',
});
/**
 * TransitGatewayRouteTableAssociation construct test
 */
describe('TransitGatewayRouteTableAssociation', () => {
  /**
   * Snapshot test
   */
  test(`${testNamePrefix} Snapshot Test`, () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
  /**
   * Number of TransitGatewayRouteTableAssociation resource test
   */
  test(`${testNamePrefix} TransitGatewayRouteTableAssociation resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::EC2::TransitGatewayRouteTableAssociation', 1);
  });

  /**
   * TransitGatewayRouteTableAssociation resource configuration test
   */
  test(`${testNamePrefix} TransitGatewayRouteTableAssociation resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        TransitGatewayRouteTableAssociation19E386E4: {
          Type: 'AWS::EC2::TransitGatewayRouteTableAssociation',
          Properties: {
            TransitGatewayAttachmentId: 'transitGatewayAttachmentId',
            TransitGatewayRouteTableId: 'transitGatewayRouteTableId',
          },
        },
      },
    });
  });
});
