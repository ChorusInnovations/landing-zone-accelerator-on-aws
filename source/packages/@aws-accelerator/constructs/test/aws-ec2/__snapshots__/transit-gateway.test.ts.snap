// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitGateway Construct(TransitGateway):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354": Object {
      "DependsOn": Array [
        "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomGetTransitGatewayAttachmentCustomResourceProviderLogGroup41699CF3": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KmsKeyId": Object {
          "Fn::GetAtt": Array [
            "TgwAttachLookupKms4975C754",
            "Arn",
          ],
        },
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "sts:AssumeRole",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "ec2:DescribeTransitGatewayAttachments",
                    "ec2:DescribeVpnConnections",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TgwAttachLookupC88D0DA8": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CustomGetTransitGatewayAttachmentCustomResourceProviderLogGroup41699CF3",
      ],
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354",
            "Arn",
          ],
        },
        "name": "name",
        "region": Object {
          "Ref": "AWS::Region",
        },
        "roleArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":iam::owningAccountId:role/roleName",
            ],
          ],
        },
        "transitGatewayId": "transitGatewayId",
        "type": "vpc",
        "uuid": "REPLACED-UUID",
      },
      "Type": "Custom::GetTransitGatewayAttachment",
      "UpdateReplacePolicy": "Delete",
    },
    "TgwAttachLookupKms4975C754": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TransitGateway888C3350": Object {
      "Properties": Object {
        "AmazonSideAsn": 1234,
        "AutoAcceptSharedAttachments": "enable",
        "DefaultRouteTableAssociation": "enable",
        "DefaultRouteTablePropagation": "enable",
        "DnsSupport": "enable",
        "Tags": Array [
          Object {
            "Key": "key",
            "Value": "value",
          },
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "VpnEcmpSupport": "enable",
      },
      "Type": "AWS::EC2::TransitGateway",
    },
    "TransitGatewayAttachment610A3027": Object {
      "Properties": Object {
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
    "TransitGatewayAttachmentAwsPartition24B0B378": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "enable",
          "DnsSupport": "disable",
          "Ipv6Support": "enable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayVpcAttachment",
    },
    "TransitGatewayAttachmentAwsPartitionOptions0EBDAD23": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "disable",
          "DnsSupport": "enable",
          "Ipv6Support": "disable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayVpcAttachment",
    },
    "TransitGatewayAttachmentGovCloud09E6C5F1": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "enable",
          "DnsSupport": "enable",
          "Ipv6Support": "disable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
    "TransitGatewayRouteTableAssociation19E386E4": Object {
      "Properties": Object {
        "TransitGatewayAttachmentId": "transitGatewayAttachmentId",
        "TransitGatewayRouteTableId": "transitGatewayRouteTableId",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTableAssociation",
    },
    "TransitGatewayRouteTablePropagationBC0F0909": Object {
      "Properties": Object {
        "TransitGatewayAttachmentId": "transitGatewayAttachmentId",
        "TransitGatewayRouteTableId": "transitGatewayRouteTableId",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTablePropagation",
    },
  },
}
`;

exports[`TransitGatewayAttachment Construct(TransitGatewayAttachment):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354": Object {
      "DependsOn": Array [
        "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomGetTransitGatewayAttachmentCustomResourceProviderLogGroup41699CF3": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KmsKeyId": Object {
          "Fn::GetAtt": Array [
            "TgwAttachLookupKms4975C754",
            "Arn",
          ],
        },
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "sts:AssumeRole",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "ec2:DescribeTransitGatewayAttachments",
                    "ec2:DescribeVpnConnections",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TgwAttachLookupC88D0DA8": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CustomGetTransitGatewayAttachmentCustomResourceProviderLogGroup41699CF3",
      ],
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354",
            "Arn",
          ],
        },
        "name": "name",
        "region": Object {
          "Ref": "AWS::Region",
        },
        "roleArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":iam::owningAccountId:role/roleName",
            ],
          ],
        },
        "transitGatewayId": "transitGatewayId",
        "type": "vpc",
        "uuid": "REPLACED-UUID",
      },
      "Type": "Custom::GetTransitGatewayAttachment",
      "UpdateReplacePolicy": "Delete",
    },
    "TgwAttachLookupKms4975C754": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TransitGateway888C3350": Object {
      "Properties": Object {
        "AmazonSideAsn": 1234,
        "AutoAcceptSharedAttachments": "enable",
        "DefaultRouteTableAssociation": "enable",
        "DefaultRouteTablePropagation": "enable",
        "DnsSupport": "enable",
        "Tags": Array [
          Object {
            "Key": "key",
            "Value": "value",
          },
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "VpnEcmpSupport": "enable",
      },
      "Type": "AWS::EC2::TransitGateway",
    },
    "TransitGatewayAttachment610A3027": Object {
      "Properties": Object {
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
    "TransitGatewayAttachmentAwsPartition24B0B378": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "enable",
          "DnsSupport": "disable",
          "Ipv6Support": "enable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayVpcAttachment",
    },
    "TransitGatewayAttachmentAwsPartitionOptions0EBDAD23": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "disable",
          "DnsSupport": "enable",
          "Ipv6Support": "disable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayVpcAttachment",
    },
    "TransitGatewayAttachmentGovCloud09E6C5F1": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "enable",
          "DnsSupport": "enable",
          "Ipv6Support": "disable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
    "TransitGatewayRouteTableAssociation19E386E4": Object {
      "Properties": Object {
        "TransitGatewayAttachmentId": "transitGatewayAttachmentId",
        "TransitGatewayRouteTableId": "transitGatewayRouteTableId",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTableAssociation",
    },
    "TransitGatewayRouteTablePropagationBC0F0909": Object {
      "Properties": Object {
        "TransitGatewayAttachmentId": "transitGatewayAttachmentId",
        "TransitGatewayRouteTableId": "transitGatewayRouteTableId",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTablePropagation",
    },
  },
}
`;

exports[`TransitGatewayRouteTableAssociation Construct(TransitGatewayRouteTableAssociation):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "TransitGateway888C3350": Object {
      "Properties": Object {
        "AmazonSideAsn": 1234,
        "AutoAcceptSharedAttachments": "enable",
        "DefaultRouteTableAssociation": "enable",
        "DefaultRouteTablePropagation": "enable",
        "DnsSupport": "enable",
        "Tags": Array [
          Object {
            "Key": "key",
            "Value": "value",
          },
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "VpnEcmpSupport": "enable",
      },
      "Type": "AWS::EC2::TransitGateway",
    },
    "TransitGatewayRouteTableAssociation19E386E4": Object {
      "Properties": Object {
        "TransitGatewayAttachmentId": "transitGatewayAttachmentId",
        "TransitGatewayRouteTableId": "transitGatewayRouteTableId",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTableAssociation",
    },
    "TransitGatewayRouteTablePropagationBC0F0909": Object {
      "Properties": Object {
        "TransitGatewayAttachmentId": "transitGatewayAttachmentId",
        "TransitGatewayRouteTableId": "transitGatewayRouteTableId",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTablePropagation",
    },
  },
}
`;

exports[`TransitGatewayRouteTablePropagation Construct(TransitGatewayRouteTablePropagation):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354": Object {
      "DependsOn": Array [
        "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomGetTransitGatewayAttachmentCustomResourceProviderLogGroup41699CF3": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KmsKeyId": Object {
          "Fn::GetAtt": Array [
            "TgwAttachLookupKms4975C754",
            "Arn",
          ],
        },
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomGetTransitGatewayAttachmentCustomResourceProviderRoleA6A22C3D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "sts:AssumeRole",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "ec2:DescribeTransitGatewayAttachments",
                    "ec2:DescribeVpnConnections",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TgwAttachLookupC88D0DA8": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CustomGetTransitGatewayAttachmentCustomResourceProviderLogGroup41699CF3",
      ],
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomGetTransitGatewayAttachmentCustomResourceProviderHandler7E079354",
            "Arn",
          ],
        },
        "name": "name",
        "region": Object {
          "Ref": "AWS::Region",
        },
        "roleArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":iam::owningAccountId:role/roleName",
            ],
          ],
        },
        "transitGatewayId": "transitGatewayId",
        "type": "vpc",
        "uuid": "REPLACED-UUID",
      },
      "Type": "Custom::GetTransitGatewayAttachment",
      "UpdateReplacePolicy": "Delete",
    },
    "TgwAttachLookupKms4975C754": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TransitGateway888C3350": Object {
      "Properties": Object {
        "AmazonSideAsn": 1234,
        "AutoAcceptSharedAttachments": "enable",
        "DefaultRouteTableAssociation": "enable",
        "DefaultRouteTablePropagation": "enable",
        "DnsSupport": "enable",
        "Tags": Array [
          Object {
            "Key": "key",
            "Value": "value",
          },
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "VpnEcmpSupport": "enable",
      },
      "Type": "AWS::EC2::TransitGateway",
    },
    "TransitGatewayAttachment610A3027": Object {
      "Properties": Object {
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
    "TransitGatewayAttachmentAwsPartition24B0B378": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "enable",
          "DnsSupport": "disable",
          "Ipv6Support": "enable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayVpcAttachment",
    },
    "TransitGatewayAttachmentAwsPartitionOptions0EBDAD23": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "disable",
          "DnsSupport": "enable",
          "Ipv6Support": "disable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayVpcAttachment",
    },
    "TransitGatewayAttachmentGovCloud09E6C5F1": Object {
      "Properties": Object {
        "Options": Object {
          "ApplianceModeSupport": "enable",
          "DnsSupport": "enable",
          "Ipv6Support": "disable",
        },
        "SubnetIds": Array [
          "one",
          "two",
          "three",
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "name",
          },
        ],
        "TransitGatewayId": "transitGatewayId",
        "VpcId": "vpcId",
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
    "TransitGatewayRouteTableAssociation19E386E4": Object {
      "Properties": Object {
        "TransitGatewayAttachmentId": "transitGatewayAttachmentId",
        "TransitGatewayRouteTableId": "transitGatewayRouteTableId",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTableAssociation",
    },
    "TransitGatewayRouteTablePropagationBC0F0909": Object {
      "Properties": Object {
        "TransitGatewayAttachmentId": "transitGatewayAttachmentId",
        "TransitGatewayRouteTableId": "transitGatewayRouteTableId",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTablePropagation",
    },
  },
}
`;
